File: Providers/STSimMapProvider.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports System.Reflection
Imports System.Globalization
Imports SyncroSim.Core
Imports SyncroSim.StochasticTime.Forms

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class STSimMapProvider
    Inherits MapProvider

    Public Overrides Sub CreateColorMaps(project As Project)

        'STATECLASS Color Map and Legend Map
        Dim dicLegendColors = CreateLegendMap(project, SPATIAL_MAP_STATE_CLASS_VARIABLE_NAME, DATASHEET_STATECLASS_NAME)
        CreateColorMap(project, SPATIAL_MAP_STATE_CLASS_VARIABLE_NAME, DATASHEET_STATECLASS_NAME, dicLegendColors)

        'Primary Stratum Color Map and Legend Map
        dicLegendColors = CreateLegendMap(project, SPATIAL_MAP_STRATUM_VARIABLE_NAME, DATASHEET_STRATA_NAME)
        CreateColorMap(project, SPATIAL_MAP_STRATUM_VARIABLE_NAME, DATASHEET_STRATA_NAME, dicLegendColors)

        'Transition Groups Color Map and Legend Map
        CreateTransitionGroupMaps(project)

        'Age  Color Map 
        CreateAgeColorMap(project)

    End Sub

    Public Overrides Sub RefreshCriteria(layout As SyncroSimLayout, project As Project)

        Using store As DataStore = project.Library.CreateDataStore

            Dim g0 As New SyncroSimLayoutItem("BasicGroup", "Basic", True)
            Dim g1 As New SyncroSimLayoutItem("TransitionsGroup", "Transitions", True)
            Dim g2 As New SyncroSimLayoutItem("AATPGroup", "Average Annual Probability", True)
            Dim g3 As New SyncroSimLayoutItem("StateAttributeGroup", "State Attributes", True)
            Dim g4 As New SyncroSimLayoutItem("TransitionAttributeGroup", "Transition Attributes", True)

            Dim AttrGroupView As DataView = CreateMapAttributeGroupsView(project, store)

            'Basic
            AddBasicVariables(project, g0)

            'Transitions
            AddMapTransitionGroupVariables(project, store, g1.Items,
                "STSim_OutputSpatialTransition", "Filename", "TransitionGroupID", "(Transitions)",
                SPATIAL_MAP_TRANSITION_GROUP_VARIABLE_PREFIX, False)

            'Average Annual Transition Probability
            AddMapTransitionGroupVariables(project, store, g2.Items,
                "STSim_OutputSpatialAverageTransitionProbability", "Filename", "TransitionGroupID", "(Avg. Annual Prob. - All Iterations)",
                SPATIAL_MAP_AVG_ANNUAL_TRANSITION_PROBABILITY_VARIABLE_PREFIX, True)

            'State Attributes
            AddMapStateAttributes(g3.Items, project, store, AttrGroupView)

            'Transition Attributes
            AddMapTransitionAttributes(g4.Items, project, store, AttrGroupView)

            layout.Items.Add(g0)

            If (g1.Items.Count > 0) Then
                layout.Items.Add(g1)
            End If

            If (g2.Items.Count > 0) Then
                layout.Items.Add(g2)
            End If

            If (g3.Items.Count > 0) Then
                layout.Items.Add(g3)
            End If

            If (g4.Items.Count > 0) Then
                layout.Items.Add(g4)
            End If

        End Using

    End Sub

    Private Shared Function CreateMapAttributeGroupsView(ByVal project As Project, ByVal store As DataStore) As DataView

        Dim ds As DataSheet = project.GetDataSheet(DATASHEET_ATTRIBUTE_GROUP_NAME)

        Dim View As New DataView(
            ds.GetData(store),
            Nothing,
            ds.ValidationTable.DisplayMember,
            DataViewRowState.CurrentRows)

        Return View

    End Function

    Private Shared Sub AddBasicVariables(ByVal project As Project, ByVal g0 As SyncroSimLayoutItem)

        Dim psl As String = Nothing
        Dim ssl As String = Nothing
        Dim dsterm As DataSheet = project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        GetStratumLabelTerminology(dsterm, psl, ssl)

        Dim i1 As SyncroSimLayoutItem = New SyncroSimLayoutItem(SPATIAL_MAP_STATE_CLASS_VARIABLE_NAME, "State Class", False)
        Dim i2 As SyncroSimLayoutItem = New SyncroSimLayoutItem(SPATIAL_MAP_AGE_VARIABLE_NAME, "Age", False)
        Dim i3 As SyncroSimLayoutItem = New SyncroSimLayoutItem(SPATIAL_MAP_STRATUM_VARIABLE_NAME, psl, False)

        i1.Properties.Add(New MetaDataProperty("dataSheet", "STSim_OutputSpatialState"))
        i1.Properties.Add(New MetaDataProperty("column", "Filename"))

        i2.Properties.Add(New MetaDataProperty("dataSheet", "STSim_OutputSpatialAge"))
        i2.Properties.Add(New MetaDataProperty("column", "Filename"))

        i3.Properties.Add(New MetaDataProperty("dataSheet", "STSim_OutputSpatialStratum"))
        i3.Properties.Add(New MetaDataProperty("column", "Filename"))

        g0.Items.Add(i1)
        g0.Items.Add(i2)
        g0.Items.Add(i3)

    End Sub

    Private Shared Sub AddMapTransitionGroupVariables(
        ByVal project As Project,
        ByVal store As DataStore,
        ByVal items As SyncroSimLayoutItemCollection,
        ByVal dataSheetName As String,
        ByVal fileColumnName As String,
        ByVal filterColumnName As String,
        ByVal extendedIdentifier As String,
        ByVal prefix As String,
        ByVal userDefinedColorMap As Boolean)

        Dim dstg As DataSheet = project.GetDataSheet(DATASHEET_TRANSITION_GROUP_NAME)
        Dim dsttg As DataSheet = project.GetDataSheet(DATASHEET_TRANSITION_TYPE_GROUP_NAME)
        Dim dttg As DataTable = dstg.GetData(store)
        Dim dtttg As DataTable = dsttg.GetData(store)

        Dim dvtg As New DataView(
            dttg,
            Nothing,
            dstg.ValidationTable.DisplayMember,
            DataViewRowState.CurrentRows)

        For Each drv As DataRowView In dvtg

            Dim tgid As Integer = CInt(drv.Row(dstg.ValidationTable.ValueMember))
            Dim query As String = String.Format(CultureInfo.InvariantCulture, "{0}={1}", DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME, tgid)
            Dim rows() As DataRow = dtttg.Select(query)

            For Each dr As DataRow In rows

                If (IsPrimaryTypeByGroup(dr)) Then

                    Dim DisplayName As String = CStr(drv.Row(dstg.ValidationTable.DisplayMember))
                    Dim VarName As String = String.Format(CultureInfo.InvariantCulture, "{0}-{1}", prefix, tgid)

                    Dim Item As New SyncroSimLayoutItem(VarName, DisplayName, False)

                    Item.Properties.Add(New MetaDataProperty("dataSheet", dataSheetName))
                    Item.Properties.Add(New MetaDataProperty("column", fileColumnName))
                    Item.Properties.Add(New MetaDataProperty("filter", filterColumnName))
                    Item.Properties.Add(New MetaDataProperty("extendedIdentifier", extendedIdentifier))
                    Item.Properties.Add(New MetaDataProperty("itemId", tgid.ToString(CultureInfo.InvariantCulture)))
                    Item.Properties.Add(New MetaDataProperty("userDefinedColorMap", userDefinedColorMap.ToString(CultureInfo.InvariantCulture)))

                    items.Add(Item)

                    Exit For

                End If

            Next

        Next

    End Sub

    Private Shared Sub AddMapStateAttributes(
        ByVal items As SyncroSimLayoutItemCollection,
        ByVal project As Project,
        ByVal store As DataStore,
        ByVal attrGroupsView As DataView)

        Dim StateAttrsDataSheet As DataSheet =
            project.GetDataSheet(DATASHEET_STATE_ATTRIBUTE_TYPE_NAME)

        AddMapNonGroupedAttributes(
            store,
            items,
            StateAttrsDataSheet,
            "STSim_OutputSpatialStateAttribute",
            "Filename",
            "StateAttributeTypeID",
            SPATIAL_MAP_STATE_ATTRIBUTE_VARIABLE_PREFIX)

        Dim GroupsDict As New Dictionary(Of String, SyncroSimLayoutItem)
        Dim GroupsList As New List(Of SyncroSimLayoutItem)

        For Each drv As DataRowView In attrGroupsView

            Dim GroupName As String = CStr(drv.Row(DATASHEET_NAME_COLUMN_NAME))
            Dim Group As New SyncroSimLayoutItem(GroupName, GroupName, True)

            GroupsDict.Add(GroupName, Group)
            GroupsList.Add(Group)

        Next

        AddMapGroupedAttributes(
            store,
            GroupsDict,
            StateAttrsDataSheet,
            "STSim_OutputSpatialStateAttribute",
            "Filename",
            "StateAttributeTypeID",
            SPATIAL_MAP_STATE_ATTRIBUTE_VARIABLE_PREFIX)

        For Each g As SyncroSimLayoutItem In GroupsList

            If (g.Items.Count > 0) Then
                items.Add(g)
            End If

        Next

    End Sub

    Private Shared Sub AddMapTransitionAttributes(
        ByVal items As SyncroSimLayoutItemCollection,
        ByVal project As Project,
        ByVal store As DataStore,
        ByVal attrGroupsView As DataView)

        Dim TransitionAttrsDataSheet As DataSheet =
            project.GetDataSheet(DATASHEET_TRANSITION_ATTRIBUTE_TYPE_NAME)

        AddMapNonGroupedAttributes(
            store,
            items,
            TransitionAttrsDataSheet,
            "STSim_OutputSpatialTransitionAttribute",
            "Filename",
            "TransitionAttributeTypeID",
            SPATIAL_MAP_TRANSITION_ATTRIBUTE_VARIABLE_PREFIX)

        Dim GroupsDict As New Dictionary(Of String, SyncroSimLayoutItem)
        Dim GroupsList As New List(Of SyncroSimLayoutItem)

        For Each drv As DataRowView In attrGroupsView

            Dim GroupName As String = CStr(drv.Row(DATASHEET_NAME_COLUMN_NAME))
            Dim Group As New SyncroSimLayoutItem(GroupName, GroupName, True)

            GroupsDict.Add(GroupName, Group)
            GroupsList.Add(Group)

        Next

        AddMapGroupedAttributes(
            store,
            GroupsDict,
            TransitionAttrsDataSheet,
            "STSim_OutputSpatialTransitionAttribute",
            "Filename",
            "TransitionAttributeTypeID",
            SPATIAL_MAP_TRANSITION_ATTRIBUTE_VARIABLE_PREFIX)

        For Each g As SyncroSimLayoutItem In GroupsList

            If (g.Items.Count > 0) Then
                items.Add(g)
            End If

        Next

    End Sub

    Private Shared Sub AddMapNonGroupedAttributes(
        ByVal store As DataStore,
        ByVal items As SyncroSimLayoutItemCollection,
        ByVal attrsDataSheet As DataSheet,
        ByVal dataSheetName As String,
        ByVal fileColumnName As String,
        ByVal filterColumnName As String,
        ByVal prefix As String)

        Dim Table As DataTable = attrsDataSheet.GetData(store)

        Dim View As New DataView(
            Table,
            Nothing,
            attrsDataSheet.ValidationTable.DisplayMember,
            DataViewRowState.CurrentRows)

        Debug.Assert(
            DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME =
            DATASHEET_TRANSITION_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME)

        For Each drv As DataRowView In View

            If (drv.Row(DATASHEET_ATTRIBUTE_GROUP_ID_COLUMN_NAME) Is DBNull.Value) Then

                Dim AttrId As Integer = CInt(drv.Row(attrsDataSheet.ValueMember))
                Dim AttrName As String = String.Format(CultureInfo.InvariantCulture, "{0}-{1}", prefix, AttrId)
                Dim DisplayName As String = CStr(drv.Row(attrsDataSheet.ValidationTable.DisplayMember))

                If (drv.Row(DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME) IsNot DBNull.Value) Then
                    DisplayName = String.Format(CultureInfo.InvariantCulture, "{0} ({1})", DisplayName, CStr(drv.Row(DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME)))
                End If

                Dim Item As New SyncroSimLayoutItem(AttrName, DisplayName, False)

                Item.Properties.Add(New MetaDataProperty("dataSheet", dataSheetName))
                Item.Properties.Add(New MetaDataProperty("column", fileColumnName))
                Item.Properties.Add(New MetaDataProperty("filter", filterColumnName))
                Item.Properties.Add(New MetaDataProperty("itemId", AttrId.ToString(CultureInfo.InvariantCulture)))
                Item.Properties.Add(New MetaDataProperty("userDefinedColorMap", "True"))

                items.Add(Item)

            End If

        Next

    End Sub

    Private Shared Sub AddMapGroupedAttributes(
        ByVal store As DataStore,
        ByVal groupsDict As Dictionary(Of String, SyncroSimLayoutItem),
        ByVal attrsDataSheet As DataSheet,
        ByVal dataSheetName As String,
        ByVal fileColumnName As String,
        ByVal filterColumnName As String,
        ByVal prefix As String)

        Dim Table As DataTable = attrsDataSheet.GetData(store)

        Dim View As New DataView(
            Table,
            Nothing,
            attrsDataSheet.ValidationTable.DisplayMember,
            DataViewRowState.CurrentRows)

        Dim GroupsDataSheet As DataSheet =
            attrsDataSheet.Project.GetDataSheet(DATASHEET_ATTRIBUTE_GROUP_NAME)

        Debug.Assert(
            DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME =
            DATASHEET_TRANSITION_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME)

        For Each drv As DataRowView In View

            If (drv.Row(DATASHEET_ATTRIBUTE_GROUP_ID_COLUMN_NAME) IsNot DBNull.Value) Then

                Dim GroupId As Integer = CType(drv.Row(DATASHEET_ATTRIBUTE_GROUP_ID_COLUMN_NAME), Integer)
                Dim GroupName As String = GroupsDataSheet.ValidationTable.GetDisplayName(GroupId)
                Dim AttrId As Integer = CInt(drv.Row(attrsDataSheet.ValueMember))
                Dim AttrName As String = String.Format(CultureInfo.InvariantCulture, "{0}-{1}", prefix, AttrId)
                Dim DisplayName As String = CStr(drv.Row(attrsDataSheet.ValidationTable.DisplayMember))

                If (drv.Row(DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME) IsNot DBNull.Value) Then
                    DisplayName = String.Format(CultureInfo.InvariantCulture, "{0} ({1})", DisplayName, CStr(drv.Row(DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME)))
                End If

                Dim Item As New SyncroSimLayoutItem(AttrName, DisplayName, False)

                Item.Properties.Add(New MetaDataProperty("dataSheet", dataSheetName))
                Item.Properties.Add(New MetaDataProperty("column", fileColumnName))
                Item.Properties.Add(New MetaDataProperty("filter", filterColumnName))
                Item.Properties.Add(New MetaDataProperty("itemId", AttrId.ToString(CultureInfo.InvariantCulture)))
                Item.Properties.Add(New MetaDataProperty("userDefinedColorMap", "True"))

                groupsDict(GroupName).Items.Add(Item)

            End If

        Next

    End Sub

End Class

File: Providers/STSimMapProvider.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using System.Diagnostics;
using System.Globalization;
using System.Collections.Generic;
using SyncroSim.Core;
using SyncroSim.StochasticTime.Forms;

namespace SyncroSim.STSim
{
    internal class STSimMapProvider : MapProvider
    {
        public override void CreateColorMaps(Project project)
        {
            //STATECLASS Color Map and Legend Map
            var LegendColors = SpatialUtilities.CreateLegendMap(project, Constants.SPATIAL_MAP_STATE_CLASS_VARIABLE_NAME, Strings.DATASHEET_STATECLASS_NAME);
            SpatialUtilities.CreateColorMap(project, Constants.SPATIAL_MAP_STATE_CLASS_VARIABLE_NAME, Strings.DATASHEET_STATECLASS_NAME, LegendColors);

            //Primary Stratum Color Map and Legend Map
            LegendColors = SpatialUtilities.CreateLegendMap(project, Constants.SPATIAL_MAP_STRATUM_VARIABLE_NAME, Strings.DATASHEET_STRATA_NAME);
            SpatialUtilities.CreateColorMap(project, Constants.SPATIAL_MAP_STRATUM_VARIABLE_NAME, Strings.DATASHEET_STRATA_NAME, LegendColors);

            //Transition Groups Color Map and Legend Map
            SpatialUtilities.CreateTransitionGroupMaps(project);

            //Age  Color Map 
            SpatialUtilities.CreateAgeColorMap(project);
        }

        public override void RefreshCriteria(SyncroSimLayout layout, Project project)
        {
            using (DataStore store = project.Library.CreateDataStore())
            {
                SyncroSimLayoutItem g0 = new SyncroSimLayoutItem("BasicGroup", "Basic", true);
                SyncroSimLayoutItem g1 = new SyncroSimLayoutItem("TransitionsGroup", "Transitions", true);
                SyncroSimLayoutItem g2 = new SyncroSimLayoutItem("AATPGroup", "Average Annual Probability", true);
                SyncroSimLayoutItem g3 = new SyncroSimLayoutItem("StateAttributeGroup", "State Attributes", true);
                SyncroSimLayoutItem g4 = new SyncroSimLayoutItem("TransitionAttributeGroup", "Transition Attributes", true);

                DataView AttrGroupView = CreateMapAttributeGroupsView(project, store);

                //Categorical
                AddCategoricalVariables(project, g0);

                //Transitions
                AddMapTransitionGroupVariables(project, g1.Items, "STSim_OutputSpatialTransition", "Filename", "TransitionGroupID", "(Transitions)", Constants.SPATIAL_MAP_TRANSITION_GROUP_VARIABLE_PREFIX, Strings.DATASHEET_TRANSITION_TYPE_NAME);

                //Average Annual Transition Probability
                AddMapTransitionGroupVariables(project, g2.Items, "STSim_OutputSpatialAverageTransitionProbability", "Filename", "TransitionGroupID", "(Avg. Annual Prob. - All Iterations)", Constants.SPATIAL_MAP_AVG_ANNUAL_TRANSITION_PROBABILITY_VARIABLE_PREFIX, null);

                //State Attributes
                AddMapStateAttributes(g3.Items, project, store, AttrGroupView);

                //Transition Attributes
                AddMapTransitionAttributes(g4.Items, project, store, AttrGroupView);

                layout.Items.Add(g0);

                if (g1.Items.Count > 0)
                {
                    layout.Items.Add(g1);
                }

                if (g2.Items.Count > 0)
                {
                    layout.Items.Add(g2);
                }

                if (g3.Items.Count > 0)
                {
                    layout.Items.Add(g3);
                }

                if (g4.Items.Count > 0)
                {
                    layout.Items.Add(g4);
                }
            }
        }

        private static DataView CreateMapAttributeGroupsView(Project project, DataStore store)
        {
            DataSheet ds = project.GetDataSheet(Strings.DATASHEET_ATTRIBUTE_GROUP_NAME);

            DataView View = new DataView(ds.GetData(store), null, ds.ValidationTable.DisplayMember, DataViewRowState.CurrentRows);

            return View;
        }

        private static void AddCategoricalVariables(Project project, SyncroSimLayoutItem g0)
        {
            string psl = null;
            string ssl = null;
            string tsl = null;
            DataSheet dsterm = project.GetDataSheet(Strings.DATASHEET_TERMINOLOGY_NAME);
            TerminologyUtilities.GetStratumLabelTerminology(dsterm, ref psl, ref ssl, ref tsl);

            SyncroSimLayoutItem i1 = new SyncroSimLayoutItem(Constants.SPATIAL_MAP_STATE_CLASS_VARIABLE_NAME, "State Class", false);
            SyncroSimLayoutItem i2 = new SyncroSimLayoutItem(Constants.SPATIAL_MAP_AGE_VARIABLE_NAME, "Age", false);
            SyncroSimLayoutItem i3 = new SyncroSimLayoutItem(Constants.SPATIAL_MAP_STRATUM_VARIABLE_NAME, psl, false);

            i1.Properties.Add(new MetaDataProperty("dataSheet", "STSim_OutputSpatialState"));
            i1.Properties.Add(new MetaDataProperty("column", "Filename"));
            i1.Properties.Add(new MetaDataProperty("colorMapSource", Strings.DATASHEET_STATECLASS_NAME));

            i2.Properties.Add(new MetaDataProperty("dataSheet", "STSim_OutputSpatialAge"));
            i2.Properties.Add(new MetaDataProperty("column", "Filename"));
            i2.Properties.Add(new MetaDataProperty("colorMapSource", Strings.DATASHEET_AGE_GROUP_NAME));

            i3.Properties.Add(new MetaDataProperty("dataSheet", "STSim_OutputSpatialStratum"));
            i3.Properties.Add(new MetaDataProperty("column", "Filename"));
            i3.Properties.Add(new MetaDataProperty("colorMapSource", Strings.DATASHEET_STRATA_NAME));

            g0.Items.Add(i1);
            g0.Items.Add(i2);
            g0.Items.Add(i3);
        }

        private static void AddMapTransitionGroupVariables(Project project, SyncroSimLayoutItemCollection items, string dataSheetName, string fileColumnName, string filterColumnName, string extendedIdentifier, string prefix, string colorMapSource)
        {
            Dictionary<int, TransitionGroup> PrimaryGroups = GetPrimaryTransitionGroups(project);
            List<TransitionGroup> PrimaryGroupList = new List<TransitionGroup>();

            foreach (TransitionGroup g in PrimaryGroups.Values)
            {
                PrimaryGroupList.Add(g);
            }

            PrimaryGroupList.Sort((TransitionGroup g1, TransitionGroup g2) =>
            {
                return string.Compare(g1.DisplayName, g2.DisplayName, StringComparison.CurrentCulture);
            });

            foreach (TransitionGroup g in PrimaryGroupList)
            {
                string VarName = string.Format(CultureInfo.InvariantCulture, "{0}-{1}", prefix, g.TransitionGroupId);
                SyncroSimLayoutItem Item = new SyncroSimLayoutItem(VarName, g.DisplayName, false);

                Item.Properties.Add(new MetaDataProperty("dataSheet", dataSheetName));
                Item.Properties.Add(new MetaDataProperty("column", fileColumnName));
                Item.Properties.Add(new MetaDataProperty("filter", filterColumnName));
                Item.Properties.Add(new MetaDataProperty("extendedIdentifier", extendedIdentifier));
                Item.Properties.Add(new MetaDataProperty("itemId", g.TransitionGroupId.ToString(CultureInfo.InvariantCulture)));
                Item.Properties.Add(new MetaDataProperty("colorMapSource", colorMapSource));

                items.Add(Item);
            }
        }

        private static Dictionary<int, TransitionGroup> GetPrimaryTransitionGroups(Project project)
        {
            TransitionTypeCollection TransitionTypes = GetTransitionTypes(project);
            TransitionGroupCollection TransitionGroups = GetTransitionGroups(project);
            Dictionary<int, bool> TransitionSimulationGroups = GetTransitionSimulationGroups(project);
            DataSheet TypeGroupDataSheet = project.GetDataSheet(Strings.DATASHEET_TRANSITION_TYPE_GROUP_NAME);
            Dictionary<int, TransitionGroup> PrimaryGroups = new Dictionary<int, TransitionGroup>();

            foreach (TransitionType TType in TransitionTypes)
            {
                string query = string.Format(CultureInfo.InvariantCulture,
                    "{0}={1}", Strings.DATASHEET_TRANSITION_TYPE_ID_COLUMN_NAME, TType.TransitionTypeId);

                DataRow[] TGroupRows = TypeGroupDataSheet.GetData().Select(query);
                bool TypeHasNonAutoPrimaryGroup = false;

                foreach (DataRow TGroupRow in TGroupRows)
                {
                    int tgid = Convert.ToInt32(TGroupRow[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                    TransitionGroup tg = TransitionGroups[tgid];

                    if (!tg.IsAuto)
                    {
                        if (TransitionSimulationGroups.ContainsKey(tgid))
                        {
                            TypeHasNonAutoPrimaryGroup = true;

                            if (!PrimaryGroups.ContainsKey(tgid))
                            {
                                PrimaryGroups.Add(tgid, tg);
                            }
                        }
                    }
                }

                if (!TypeHasNonAutoPrimaryGroup)
                {
                    foreach (DataRow TGroupRow in TGroupRows)
                    {
                        int tgid = Convert.ToInt32(TGroupRow[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                        TransitionGroup tg = TransitionGroups[tgid];

                        if (tg.IsAuto)
                        {
                            if (!PrimaryGroups.ContainsKey(tgid))
                            {
                                PrimaryGroups.Add(tgid, tg);
                            }
                        }
                    }
                }
            }

            return PrimaryGroups;
        }

        private static TransitionTypeCollection GetTransitionTypes(Project project)
        {
            TransitionTypeCollection types = new TransitionTypeCollection();
            DataSheet ds = project.GetDataSheet(Strings.DATASHEET_TRANSITION_TYPE_NAME);

            foreach (DataRow dr in ds.GetData().Rows)
            {
                int TransitionTypeId = Convert.ToInt32(dr[ds.PrimaryKeyColumn.Name], CultureInfo.InvariantCulture);
                string DisplayName = Convert.ToString(dr[ds.DisplayMember], CultureInfo.InvariantCulture);

                types.Add(new TransitionType(TransitionTypeId, DisplayName, 0));
            }

            return types;
        }

        private static TransitionGroupCollection GetTransitionGroups(Project project)
        {
            TransitionGroupCollection groups = new TransitionGroupCollection();
            DataSheet ds = project.GetDataSheet(Strings.DATASHEET_TRANSITION_GROUP_NAME);

            foreach (DataRow dr in ds.GetData().Rows)
            {
                int TransitionGroupId = Convert.ToInt32(dr[ds.PrimaryKeyColumn.Name], CultureInfo.InvariantCulture);
                string DisplayName = Convert.ToString(dr[ds.DisplayMember], CultureInfo.InvariantCulture);
                bool IsAuto = DataTableUtilities.GetDataBool(dr, Strings.IS_AUTO_COLUMN_NAME);

                groups.Add(new TransitionGroup(TransitionGroupId, DisplayName, IsAuto));
            }

            return groups;
        }

        private static Dictionary<int, bool> GetTransitionSimulationGroups(Project project)
        {
            Dictionary<int, bool> groups = new Dictionary<int, bool>();
            DataSheet ds = project.GetDataSheet(Strings.DATASHEET_TRANSITION_GROUP_NAME);

            foreach (DataRow dr in ds.GetData().Rows)
            {
                int TransitionGroupId = Convert.ToInt32(dr[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                groups.Add(TransitionGroupId, true);
            }

            return groups;
        }

        private static void AddMapStateAttributes(SyncroSimLayoutItemCollection items, Project project, DataStore store, DataView attrGroupsView)
        {
            DataSheet StateAttrsDataSheet = project.GetDataSheet(Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_NAME);

            AddMapNonGroupedAttributes(
                store, items, StateAttrsDataSheet, 
                "STSim_OutputSpatialStateAttribute", "Filename", "StateAttributeTypeID",
                Constants.SPATIAL_MAP_STATE_ATTRIBUTE_VARIABLE_PREFIX, null);

            Dictionary<string, SyncroSimLayoutItem> GroupsDict = new Dictionary<string, SyncroSimLayoutItem>();
            List<SyncroSimLayoutItem> GroupsList = new List<SyncroSimLayoutItem>();

            foreach (DataRowView drv in attrGroupsView)
            {
                string GroupName = Convert.ToString(drv.Row[Strings.DATASHEET_NAME_COLUMN_NAME], CultureInfo.InvariantCulture);
                SyncroSimLayoutItem Group = new SyncroSimLayoutItem(GroupName, GroupName, true);

                GroupsDict.Add(GroupName, Group);
                GroupsList.Add(Group);
            }

            AddMapGroupedAttributes(
                store, GroupsDict, StateAttrsDataSheet, 
                "STSim_OutputSpatialStateAttribute", "Filename", "StateAttributeTypeID", 
                Constants.SPATIAL_MAP_STATE_ATTRIBUTE_VARIABLE_PREFIX, null);

            foreach (SyncroSimLayoutItem g in GroupsList)
            {
                if (g.Items.Count > 0)
                {
                    items.Add(g);
                }
            }
        }

        private static void AddMapTransitionAttributes(SyncroSimLayoutItemCollection items, Project project, DataStore store, DataView attrGroupsView)
        {
            DataSheet TransitionAttrsDataSheet = project.GetDataSheet(Strings.DATASHEET_TRANSITION_ATTRIBUTE_TYPE_NAME);

            AddMapNonGroupedAttributes(
                store, items, TransitionAttrsDataSheet, 
                "STSim_OutputSpatialTransitionAttribute", "Filename", "TransitionAttributeTypeID", 
                Constants.SPATIAL_MAP_TRANSITION_ATTRIBUTE_VARIABLE_PREFIX, null);

            Dictionary<string, SyncroSimLayoutItem> GroupsDict = new Dictionary<string, SyncroSimLayoutItem>();
            List<SyncroSimLayoutItem> GroupsList = new List<SyncroSimLayoutItem>();

            foreach (DataRowView drv in attrGroupsView)
            {
                string GroupName = Convert.ToString(drv.Row[Strings.DATASHEET_NAME_COLUMN_NAME], CultureInfo.InvariantCulture);
                SyncroSimLayoutItem Group = new SyncroSimLayoutItem(GroupName, GroupName, true);

                GroupsDict.Add(GroupName, Group);
                GroupsList.Add(Group);
            }

            AddMapGroupedAttributes(
                store, GroupsDict, TransitionAttrsDataSheet, 
                "STSim_OutputSpatialTransitionAttribute", "Filename", "TransitionAttributeTypeID",
                Constants.SPATIAL_MAP_TRANSITION_ATTRIBUTE_VARIABLE_PREFIX, null);

            foreach (SyncroSimLayoutItem g in GroupsList)
            {
                if (g.Items.Count > 0)
                {
                    items.Add(g);
                }
            }
        }

        private static void AddMapNonGroupedAttributes(DataStore store, SyncroSimLayoutItemCollection items, DataSheet attrsDataSheet, string dataSheetName, string fileColumnName, string filterColumnName, string prefix, string colorMapSource)
        {
            DataTable Table = attrsDataSheet.GetData(store);
            DataView View = new DataView(Table, null, attrsDataSheet.ValidationTable.DisplayMember, DataViewRowState.CurrentRows);
            Debug.Assert(Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME == Strings.DATASHEET_TRANSITION_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME);

            foreach (DataRowView drv in View)
            {
                if (drv.Row[Strings.DATASHEET_ATTRIBUTE_GROUP_ID_COLUMN_NAME] == DBNull.Value)
                {
                    int AttrId = Convert.ToInt32(drv.Row[attrsDataSheet.ValueMember], CultureInfo.InvariantCulture);
                    string AttrName = string.Format(CultureInfo.InvariantCulture, "{0}-{1}", prefix, AttrId);
                    string DisplayName = Convert.ToString(drv.Row[attrsDataSheet.ValidationTable.DisplayMember], CultureInfo.InvariantCulture);

                    if (drv.Row[Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME] != DBNull.Value)
                    {
                        DisplayName = string.Format(CultureInfo.InvariantCulture, 
                            "{0} ({1})", 
                            DisplayName, 
                            Convert.ToString(drv.Row[Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME], CultureInfo.InvariantCulture));
                    }

                    SyncroSimLayoutItem Item = new SyncroSimLayoutItem(AttrName, DisplayName, false);

                    Item.Properties.Add(new MetaDataProperty("dataSheet", dataSheetName));
                    Item.Properties.Add(new MetaDataProperty("column", fileColumnName));
                    Item.Properties.Add(new MetaDataProperty("filter", filterColumnName));
                    Item.Properties.Add(new MetaDataProperty("itemId", AttrId.ToString(CultureInfo.InvariantCulture)));
                    Item.Properties.Add(new MetaDataProperty("colorMapSource", colorMapSource));

                    items.Add(Item);
                }
            }
        }

        private static void AddMapGroupedAttributes(DataStore store, Dictionary<string, SyncroSimLayoutItem> groupsDict, DataSheet attrsDataSheet, string dataSheetName, string fileColumnName, string filterColumnName, string prefix, string colorMapSource)
        {
            DataTable Table = attrsDataSheet.GetData(store);
            DataView View = new DataView(Table, null, attrsDataSheet.ValidationTable.DisplayMember, DataViewRowState.CurrentRows);
            DataSheet GroupsDataSheet = attrsDataSheet.Project.GetDataSheet(Strings.DATASHEET_ATTRIBUTE_GROUP_NAME);
            Debug.Assert(Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME == Strings.DATASHEET_TRANSITION_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME);

            foreach (DataRowView drv in View)
            {
                if (drv.Row[Strings.DATASHEET_ATTRIBUTE_GROUP_ID_COLUMN_NAME] != DBNull.Value)
                {
                    int GroupId = Convert.ToInt32(drv.Row[Strings.DATASHEET_ATTRIBUTE_GROUP_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                    string GroupName = GroupsDataSheet.ValidationTable.GetDisplayName(GroupId);
                    int AttrId = Convert.ToInt32(drv.Row[attrsDataSheet.ValueMember], CultureInfo.InvariantCulture);
                    string AttrName = string.Format(CultureInfo.InvariantCulture, "{0}-{1}", prefix, AttrId);
                    string DisplayName = Convert.ToString(drv.Row[attrsDataSheet.ValidationTable.DisplayMember], CultureInfo.InvariantCulture);

                    if (drv.Row[Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME] != DBNull.Value)
                    {
                        DisplayName = string.Format(CultureInfo.InvariantCulture, 
                            "{0} ({1})", 
                            DisplayName, 
                            Convert.ToString(drv.Row[Strings.DATASHEET_STATE_ATTRIBUTE_TYPE_UNITS_COLUMN_NAME], CultureInfo.InvariantCulture));
                    }

                    SyncroSimLayoutItem Item = new SyncroSimLayoutItem(AttrName, DisplayName, false);

                    Item.Properties.Add(new MetaDataProperty("dataSheet", dataSheetName));
                    Item.Properties.Add(new MetaDataProperty("column", fileColumnName));
                    Item.Properties.Add(new MetaDataProperty("filter", filterColumnName));
                    Item.Properties.Add(new MetaDataProperty("itemId", AttrId.ToString(CultureInfo.InvariantCulture)));
                    Item.Properties.Add(new MetaDataProperty("colorMapSource", colorMapSource));

                    groupsDict[GroupName].Items.Add(Item);
                }
            }
        }
    }
}
